src/server/journal/streamer.cc

```diff
@@ -11,62 +11,178 @@
+}
+
+JournalStreamer::~JournalStreamer() {
+  DCHECK_EQ(in_flight_bytes_, 0u);
+  VLOG(1) << "~JournalStreamer";
+}
+
+void JournalStreamer::Start(io::AsyncSink* dest, bool send_lsn) {
+  CHECK(dest_ == nullptr && dest != nullptr);
+  dest_ = dest;
   journal_cb_id_ =
       journal_->RegisterOnChange([this, send_lsn](const JournalItem& item, bool allow_await) {
-        if (!ShouldWrite(item)) {
-          return;
+        if (allow_await) {
+          ThrottleIfNeeded();
+          // No record to write, just await if data was written so consumer will read the data.
+          if (item.opcode == Op::NOOP)
+            return;
         }
 
-        if (item.opcode == Op::NOOP) {
-          // No record to write, just await if data was written so consumer will read the data.
-          return AwaitIfWritten();
+        if (!ShouldWrite(item)) {
+          return;
         }
 
-        Write(io::Buffer(item.data));
+        Write(item.data);
         time_t now = time(nullptr);
+
+        // TODO: to chain it to the previous Write call.
         if (send_lsn && now - last_lsn_time_ > 3) {
           last_lsn_time_ = now;
-          base::IoBuf tmp;
-          io::BufSink sink(&tmp);
+          io::StringSink sink;
           JournalWriter writer(&sink);
           writer.Write(Entry{journal::Op::LSN, item.lsn});
-          Write(io::Buffer(io::View(tmp.InputBuffer())));
+          Write(sink.str());
         }
-        NotifyWritten(allow_await);
       });
 }
 
 void JournalStreamer::Cancel() {
-  Finalize();  // Finalize must be called before UnregisterOnChange because we first need to stop
-               // writing to buffer and notify the all the producers.
-               // Writing to journal holds mutex protecting change_cb_arr_, than the fiber can
-               // preemt when calling NotifyWritten and it will not run again till notified.
-               // UnregisterOnChange will try to lock the mutex therefor calling UnregisterOnChange
-               // before Finalize may cause deadlock.
+  VLOG(1) << "JournalStreamer::Cancel";
+  is_stopped_ = true;
+  waker_.notifyAll();
   journal_->UnregisterOnChange(journal_cb_id_);
+  WaitForInflightToComplete();
+}
+
+size_t JournalStreamer::GetTotalBufferCapacities() const {
+  return in_flight_bytes_ + pending_buf_.capacity();
+}
 
-  if (write_fb_.IsJoinable()) {
-    write_fb_.Join();
+constexpr size_t kFlushThreshold = 2_KB;
+
+void JournalStreamer::Write(std::string_view str) {
+  DCHECK(!str.empty());
+  DVLOG(2) << "Writing " << str.size() << " bytes";
+
+  // If we do not have any in flight requests we send the string right a way.
+  // We can not aggregate it since we do not know when the next update will follow.
+  size_t total_pending = pending_buf_.size() + str.size();
+  if (in_flight_bytes_ == 0 || total_pending > kFlushThreshold) {
+    // because of potential SOO with strings we allocate explicitly on heap
+    uint8_t* buf(new uint8_t[str.size()]);
+
```

Comment:
we can have a swapbuffer in the meantime for v19

Reply:
I will submit it to 1.19

Reply:
what's a swapbuffer?